google/auth/pluggable.py

```diff
@@ -0,0 +1,327 @@
+            ] = original_credential_source_executable_output_file
+        elif os.getenv("GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE") is not None:
+            del os.environ["GOOGLE_EXTERNAL_ACCOUNT_OUTPUT_FILE"]
+
+        if result.returncode != 0:
+            raise exceptions.RefreshError(
+                "Executable exited with non-zero return code {}. Error: {}".format(
+                    result.returncode, result.stdout
+                )
+            )
+        else:
+            data = result.stdout.decode("utf-8")
+            response = json.loads(data)
+            return self._parse_subject_token(response)
+
+    @classmethod
+    def from_info(cls, info, **kwargs):
+        """Creates a Pluggable Credentials instance from parsed external account info.
+
+        Args:
+            info (Mapping[str, str]): The Pluggable external account info in Google
+                format.
+            kwargs: Additional arguments to pass to the constructor.
+
+        Returns:
+            google.auth.pluggable.Credentials: The constructed
+                credentials.
+
+        Raises:
+            ValueError: For invalid parameters.
+        """
+        return cls(
+            audience=info.get("audience"),
+            subject_token_type=info.get("subject_token_type"),
+            token_url=info.get("token_url"),
+            service_account_impersonation_url=info.get(
+                "service_account_impersonation_url"
+            ),
+            client_id=info.get("client_id"),
+            client_secret=info.get("client_secret"),
+            credential_source=info.get("credential_source"),
+            quota_project_id=info.get("quota_project_id"),
+            workforce_pool_user_project=info.get("workforce_pool_user_project"),
+            **kwargs
+        )
+
+    @classmethod
+    def from_file(cls, filename, **kwargs):
+        """Creates an Pluggable Credentials instance from an external account json file.
+
+        Args:
+            filename (str): The path to the Pluggable external account json file.
+            kwargs: Additional arguments to pass to the constructor.
+
+        Returns:
+            google.auth.pluggable.Credentials: The constructed
+                credentials.
+        """
+        with io.open(filename, "r", encoding="utf-8") as json_file:
+            data = json.load(json_file)
+            return cls.from_info(data, **kwargs)
+
+    def _parse_subject_token(self, response):
+        if not response["success"]:
+            raise exceptions.RefreshError(
+                "Executable returned unsuccessful response: {}.".format(response)
+            )
+        elif response["version"] > EXECUTABLE_SUPPORTED_MAX_VERSION:
+            raise exceptions.RefreshError(
+                "Executable returned unsupported version {}.".format(
+                    response["version"]
+                )
+            )
+        elif response["expiration_time"] < time.time():
```

Comment:
Do we have any errors that have a code and a message?

For instance, if expired, we can return {INVALID_RESPONSE, "The executable response is expired.")

Reply:
I don't see any errors with error code though. They're pure messages.

Reply:
Ok. In Java, I defined a new PluggableAuthException. Maybe we can do something similar here?

Reply:
Looks better, let me try add something similar.

Reply:
We throw 2 types of error here: ValueError(https://docs.python.org/3/tutorial/errors.html#raising-exceptions) and google.auth.exceptions(https://google-auth.readthedocs.io/en/stable/reference/google.auth.exceptions.html#google.auth.exceptions.RefreshError). Both of them take only a string to construct.

I tried to define an error class, however I think it makes the error construction more complicated and less clear. It also makes it hard for the system to read the string literal of the error class. WDYT?

Reply:
Let's keep it like this then if this is how it currently works in the lib.